src/core/Refiner.ml

```diff
@@ -486,6 +490,60 @@ struct
 end
 
 
+module Signature =
+struct
+  let formation (tacs : T.Tp.tac telescope) : T.Tp.tac =
+    let rec form_fields tele =
+      function
+      | Bind (nm, tac, tacs) ->
+         let* tp = T.Tp.run tac in
+         let* vtp = RM.lift_ev @@ Sem.eval_tp tp in
+         T.abstract ~ident:(`User [nm]) vtp @@ fun var -> form_fields (Snoc (tele, (nm, tp))) (tacs var)
+      | Done -> RM.ret @@ S.Signature (Bwd.to_list tele)
+    in T.Tp.rule @@ form_fields Emp tacs
+
+  let rec intro_fields phi phi_clo (sign : D.sign) (tacs : (string * T.Chk.tac) list) : (string * S.t) list m =
+    match (sign, tacs) with
+    | D.Field (lbl, tp, sign_clo), (tac_lbl, tac) :: tacs ->
+       let* _ = MU.guard (not @@ String.equal lbl tac_lbl) @@ fun () ->
+          RM.expected_field_name ~expected:lbl ~actual:tac_lbl
+       in
+       (* FIXME: Take Cofibrations into account *)
+       let* tfield = T.Chk.brun tac (tp, phi, phi_clo) in
```

Comment:
In case it helps, I think that `phi_clo` needs to be changed to project out the corresponding field. I think you can find an example of doing so in the intro rule for the legacy sigma types.

Reply:
Yeah, that makes sense! I've made the change